### PR TITLE
Eslint and Prettier integration

### LINKED ISSUE #284: Better linting configuration and build integration
#283 updates the project to use `eslint` instead of `tslint`. There are a [few more steps](https://code.visualstudio.com/api/advanced-topics/tslint-eslint-migration) required for effective linting:
- [Extend the configuration](https://github.com/typescript-eslint/tslint-to-eslint-config/blob/main/docs/FAQs.md#should-i-use-tslint-to-eslint-config) from `typescript-eslint`'s recommended rulesets.
  ```json
    "extends": [
        "eslint:recommended",
        "plugin:@typescript-eslint/recommended"
    ],
  ```
  - This will involve cleaning up the `.eslintrc.js` config file to avoid re-specifying rules.
- [Use Prettier](https://github.com/typescript-eslint/tslint-to-eslint-config/blob/main/docs/FAQs.md#should-i-use-prettier) rather than the linter to handle formatting concerns.
  - We might use [`eslint-config-prettier`](https://github.com/prettier/eslint-config-prettier) as part of this.
- Add linting to the `webview-ui` project. The lint config currently points to `tsconfig.json` as the project root, and `webview-ui` has its own TS configuration.
- Consider [typed linting](https://typescript-eslint.io/linting/typed-linting/).
- Fix up all linting warnings/errors.
- Convert warnings to errors and run linting as part of build scripts/pipelines for CI.

### PR DESCRIPTION
**Pull request by peterbom**
Resolves #284 
- Adds eslint to `webview-ui` project
- Makes rules consistent between projects
- Ensures linting of both projects is done as part of build
- Delegates formatting responsibilities to Prettier
- Runs Prettier on save in VS Code

### CODE REVIEW

File: src/commands/aksInspektorGadget/traceItems.ts
Comment (peterbom) [inline src/commands/aksInspektorGadget/traceItems.ts:40]: These are moved to `utils/runtimeTypes.ts`

File: src/commands/periscope/helpers/periscopehelper.ts
Comment (peterbom) [inline src/commands/periscope/helpers/periscopehelper.ts:105]: Apparently the previous code was "unsafe and wrong", which made me feel very ashamed of myself, so it's replaced by a bunch of very explicit checks.

File: src/commands/utils/azureAccount.ts
Comment (peterbom) [inline src/commands/utils/azureAccount.ts:225]: I don't think this will behave differently in practice, because errors are _usually_ objects (especially from trusted libraries), but this adds and extra check just to make sure.

File: src/commands/utils/config.ts
Comment (peterbom) [inline src/commands/utils/config.ts:117]: This should be equivalent to the previous code, except the old code wouldn't handle nulls, whereas `isObject` does.

File: src/commands/utils/detectors.ts
Comment (peterbom) [inline src/commands/utils/detectors.ts:12]: Not sure why we were using `require()` for importing `tmp` functions, so I thought I'd better call this out in case there's a reason I don't understand. It seems to build and run fine, so I _think_ this change is fine.

File: src/commands/utils/detectors.ts
Comment (peterbom) [inline src/commands/utils/detectors.ts:18]: Just some minor refactoring to move the nested function out.

Note, this code is not run as part of any commands, and was put in for generating realistic test data from ARM calls.

File: src/commands/utils/download/download.ts
Comment (peterbom) [inline src/commands/utils/download/download.ts:5]: I actually think this shouldn't be equivalent, but `eslint` complained that `download` was the wrong type in the previous code. I tested downloading files after the update and it still works.

File: src/commands/utils/kubectl.ts
Comment (peterbom) [inline src/commands/utils/kubectl.ts:158]: Replaced `any` cast with `asInternal` function that returns an `errorable` if the underlying implementation changes.

File: src/commands/utils/reporter.ts
Comment (peterbom) [inline src/commands/utils/reporter.ts:24]: moved this to a top-level import - not sure if there was a reason it was previously loaded dynamically?

File: src/extension.ts
Comment (peterbom) [inline src/extension.ts:48]: `context` was never used here.

File: src/panels/TcpDumpPanel.ts
Comment (peterbom) [inline src/panels/TcpDumpPanel.ts:23]: Added a second `\` character to escape the other one, which would previously have been ignored, meaning the regex would match _any_ character before the extension, not just a literal `.`. So I think the linter caught a genuine bug here.

File: src/panels/TcpDumpPanel.ts
Comment (peterbom) [inline src/panels/TcpDumpPanel.ts:36]: Escaping the `\` so that a `\` character appears in the regex string to capture a literal `-` in the command. I haven't quite thought through how this could have been correct before, so it's worth doing a bit more testing (although it does seem to work now).

File: src/tests/runTests.ts
Comment (peterbom) [inline src/tests/runTests.ts:8]: Had to change the typescript root folder from `src` to the project root, because some `.ts` files import `package.json` from the root, and so we want the compiler to be aware of that file. That means that the `out` folder now contains `src` and everything is nested one level deeper, hence the change in the relative path here.

Note, the `out` folder is only used for running tests, not building the vsix, and tests are now running successfully.

File: webview-ui/src/AzureServiceOperator/AzureServiceOperator.tsx
Comment (peterbom) [inline webview-ui/src/AzureServiceOperator/AzureServiceOperator.tsx:7]: This is changed to use a [custom hook](https://react.dev/learn/reusing-logic-with-custom-hooks). Some of the logic (like `useState`) that was in this function was not valid for an ordinary library function, but if you name the function `useXxxx` it's interpreted as a custom hook and that stuff is allowed. It also allows `useEffect`, meaning we could move the following three lines of repeated code into the function, further cleaning up the React components.

File: webview-ui/src/InspektorGadget/GadgetSelector.tsx
Comment (peterbom) [inline webview-ui/src/InspektorGadget/GadgetSelector.tsx:9]: The old type `React.HTMLAttributes<any>['className']` evaluates to `string | undefined`, so the simpler type is equivalent. (I guess the complex type name could be used by some tooling for intellisense but I don't think it provides any value for us.)

File: webview-ui/src/InspektorGadget/helpers/gadgets.ts
Comment (peterbom) [inline webview-ui/src/InspektorGadget/helpers/gadgets.ts:131]: I'm not sure whether values can ever be `null`. The previous code didn't handle `null`s, so added these extra checks to be safe.

File: webview-ui/src/Periscope/Periscope.tsx
Comment (peterbom) [inline webview-ui/src/Periscope/Periscope.tsx:15]: This removal of the empty list of dependencies (done because there were hidden dependencies) will result in more frequent upload status requests. It's worth testing this and seeing if it's too chatty.

File: webview-ui/src/utilities/runtimeTypes.ts
Comment (peterbom) [inline webview-ui/src/utilities/runtimeTypes.ts:1]: Note - this file is duplicated between the webview and extension projects. In the future it might be good to create a library of files that can be shared between the two (those that have no dependency on vscode or react). A discussion for later...

File: webview-ui/src/utilities/state.ts
Comment (peterbom) [inline webview-ui/src/utilities/state.ts:84]: This is just a type annotation and won't affect the compiled JavaScript, so there's no urgent need to resolve the TODO, but I want to keep it there as a reminder that there's something not quite right with the types here.

File: src/commands/utils/reporter.ts
Comment (peterbom) [reply src/commands/utils/reporter.ts:24]: Used by telemetry, verified that reporting still works correctly after the change.

File: src/extension.ts
Comment (peterbom) [reply src/extension.ts:48]: Used by telemetry, verified that reporting still works correctly after the change.

File: src/panels/TcpDumpPanel.ts
Comment (peterbom) [reply src/panels/TcpDumpPanel.ts:36]: Turns out that the `-` character [doesn't always need escaping](https://stackoverflow.com/a/5484178). In this case it wasn't being escaped (because the escape character `\` itself needed escaping to appear in the regex string), but it still worked because this was one of the cases where it didn't need escaping (it was correctly interpreted as a literal).

This change escapes it, which is also valid.

File: src/tests/runTests.ts
Comment (peterbom) [reply src/tests/runTests.ts:8]: Tats asked the question of whether the `tsconfig` change would increase the vsix side (because it's looking at more than just the `src` folder). It doesn't increase the size because webpack only pulls in what is needed starting from `/src/extension.ts`.

However, for performance reasons it's probably best to ensure that we're not looking at unnecessary folders for TypeScript compilation, linting, and prettifying. I'll add another update for this.

Comment (peterbom) [reply src/tests/runTests.ts:8]: Updated. `tsconfig.json` now uses `includes` (specifying `src`) instead of `excludes`. This still avoids the error when files are pulled in from outside the `src` folder.

File: webview-ui/src/Periscope/Periscope.tsx
Comment (peterbom) [reply webview-ui/src/Periscope/Periscope.tsx:15]: Manually tested, and not noticed any different behaviour.

File: .vscode/launch.json
Comment (Tatsinnit) [inline .vscode/launch.json:33]: 💡Please Note: I think we chatted about this, for wider readers - this change don't really make any functional problem with the code.

File: src/commands/aksTCPCollection/tcpDumpCollection.ts
Comment (Tatsinnit) [inline src/commands/aksTCPCollection/tcpDumpCollection.ts:11]: 💡 Please note: for wider readers: `unknown is the type-safe counterpart of any`

File: webview-ui/package.json
Comment (Tatsinnit) [inline webview-ui/package.json:8]: 💡 I think this is probably part of `nom` regular build.

**Review by Tatsinnit — APPROVED** [review]
❤️☕️🎊 Lots of progressive minor changes which were necessity, give int HUGE I see most of them are the `nit` yet - great evangelistic - slinging changes much needed for repo - awesomeness.

I have added just comment as future reader of these changes.

We should merge this, happy to do it in few hours after this approval which gives more time for anyone else to present their thoughts.

**Pull request merged**
